rfcs/003 - Multi-Region Support RFC.md

```diff
@@ -0,0 +1,149 @@
+2. Cross-schema joins may be impossible, or very awkward, for things like a DynamoDB cluster.
+3. The set of regions can't be dynamic in this approach without some very tricky changes in the engine to re-apply the "install" operations on the newly-discovered region.
+4. More mangling of TypeORM to toggle the schema the entities are in (though this is likely not too complicated).
+
+#### Regions as table prefixes
+
+Similar to the prior one, but instead of `us_west_2.security_groups` it would be `public.us_west_2_security_groups`.
+
+##### Pros
+
+1. Every table is immediately visible unlike the regions-as-schema-names approach.
+
+##### Cons
+
+1. TypeORM entity logic needs trickery to take a single logical entity and map it to several similarly-named entities, and then somehow inform the mapper of what trickery was used for which instance.
+2. You can't hide regions you don't care about, and every query has to be altered.
+
+#### `Region` mega-join table
+
+A specialized join table that contains `table`, `column`, `value`, and `region` columns to uniquely identify *any* row of any table and specify the region that applies to it. For rows that don't have a record in there, a `default_region` column in the `aws_account` would be assumed to be the answer if it is relevant. There could also be a list of `active_regions` that can trim the AWS API access to only regions the user is interested in.
+
+##### Pros
+
+1. Also requires zero schema changes to other modules, can be done in a single shot.
+2. Preserves the "default region" concept to keep inserts effectively identical.
+3. Can actually be faster by cutting out irrelevant regions.
+
+##### Cons
+
+1. Joins would be very slow and meta, with no good way to guarantee correctness as the joins will not be enforced by Postgres, but only us.
+2. Inserts for non-default regions are all two-step, and *must* be explicitly done before `iasql_apply` or it will accidentally create in the wrong region.
+3. Querying what's in what region will be much tougher than other options.
+
+#### Region-prefixed-to-@cloudId-column
+
+The IaSQL `@cloudId` concept (an ID that isn't the autogenerated database ID, but the ID the cloud provider generates once the resource is created) can double-up as the region indicator, when necessary. On insert you insert the region in question, and once it is created in the cloud, it it set to `region:cloudId` in the column.
+
+##### Pros
+
+1. This also requires no schema changes and can be updated over time.
+2. Could similarly have a `default_region` and `active_regions` concept like the mega-join table to gain its benefits.
+
+##### Cons
+
+1. Querying the records by region will be complicated with string prefix checking, so can never be performant, and also varies entity-by-entity.
+2. The cloud ID itself similarly would need to be parsed out to find the relevant part of the string, and so joins between entities *on* the cloudId become weirder.
```

Comment:
I think not all our entities have an explicit `cloudId`

Reply:
This is also true, EC2 instances don't, for instance, and we'd still have to do something weird for them. I'll add that to the cons list here.